tests/validators/string.test.ts

```diff
@@ -95,5 +95,154 @@ describe('StringValidator', () => {
+				);
+			});
+		});
+
+		describe('url', () => {
+			describe('Without any options', () => {
+				const urlPredicate = s.string.url();
+
+				test.each(['https://google.com', 'http://foo.bar'])('GIVEN %s THEN returns given value', (input) => {
+					expect(urlPredicate.parse(input)).toBe(input);
+				});
+
+				test.each(['google.com', 'foo.bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => urlPredicate.parse(input)).toThrow(new ConstraintError('s.string.url', 'Invalid URL', input, 'expected.url'));
+				});
+			});
+
+			describe('With protocol', () => {
+				const urlPredicateWithProtocol = s.string.url({ allowedProtocols: ['git:'] });
+
+				test.each(['git://foo.bar'])('GIVEN %s THEN returns given value', (input) => {
+					expect(urlPredicateWithProtocol.parse(input)).toBe(input);
+				});
+
+				test.each(['https://google.com', 'http://foo.bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => urlPredicateWithProtocol.parse(input)).toThrow(
+						new ConstraintError('s.string.url', 'Invalid URL protocol', input, 'expected.url')
+					);
+				});
+			});
+
+			describe('With domain', () => {
+				const urlPredicateWithDomain = s.string.url({ allowedDomains: ['google.com'] });
+
+				test.each(['https://google.com', 'http://google.com'])('GIVEN %s THEN returns given value', (input) => {
+					expect(urlPredicateWithDomain.parse(input)).toBe(input);
+				});
+
+				test.each(['https://foo.bar', 'http://foo.bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => urlPredicateWithDomain.parse(input)).toThrow(
+						new ConstraintError('s.string.url', 'Invalid URL domain', input, 'expected.url')
+					);
+				});
+			});
+		});
+
+		describe('uuid', () => {
+			describe('default', () => {
+				const uuidPredicate = s.string.uuid();
+
+				test.each(['450d6a23-9e6f-45d9-9d5a-fd4f6e014f16'])('GIVEN %s THEN returns given value', (input) => {
+					expect(uuidPredicate.parse(input)).toBe(input);
+				});
+
+				test.each(['6e8bc430-9a1b-4f7f-b7a5', '6e8bc430-9a1b-4f7f-b7a5'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => uuidPredicate.parse(input)).toThrow(
+						new ConstraintError('s.string.uuid', 'Invalid string format', input, 'expected.uuid')
+					);
+				});
+			});
+
+			describe('with version range', () => {
+				const uuidRangePredicate = s.string.uuid('1-4');
+
+				test.each(['6e8bc430-9a1b-4f7f-b7a5-ea4dede09a4b', '6e8bc430-9a1b-1f7f-b7a5-ea4dede09a4b', '6e8bc430-9a1b-3f7f-b7a5-ea4dede09a4b'])(
+					'GIVEN %s THEN returns given value',
+					(input) => {
+						expect(uuidRangePredicate.parse(input)).toBe(input);
+					}
+				);
+
+				test.each(['6e8bc430-9a1b-5f7f-b7a5-ea4dede09a4b'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => uuidRangePredicate.parse(input)).toThrow(
+						new ConstraintError('s.string.uuid', 'Invalid string format', input, 'expected.uuid')
```

Comment:
This error message is not clear at all...  Ideally it should say "expected UUID version to be in range ${range}" or similar

Reply:
Should I add a new param in `stringRegex` to take the `expected` message? or rn there is no way to know which `UUID` version is used.

Reply:
As long as it doesn't slow it down too much, I'd say that would be ideal

Reply:
updated